[INTERNAL][I] Add session utilities

tobous opened the pull request:
Adds a class containing static session utilities. The main benefit of
this utility class is that it can provide certain useful interactions
with the session without requiring the caller to have access to a
session object.

Currently, this class only provides methods to check whether a given
resource is shared or not.

[EDIT: This point is no longer applicable. The need for this utility class stems from the poor separation of which classes should belong to the session context and which to the plugin context. See discussion below.]
This class is currently only used in the IntelliJ part. But as it does not contains any functionality specific to IntelliJ and because it might be useful for other implementations, I added it to the core.
If there are other useful applications for such a static class holding a session object, the methods offered by this utility class can be expanded later on.

The class FileSystemChangeListener was explicitly not adjusted as it will be completely re-written in an upcoming PR anyway.

saros-infrastructure commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SessionUtils.java:
```diff
@@ -0,0 +1,82 @@
+        currentSarosSession = newSarosSession;
+    }
+
+    private static void resetSession() {
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Move this method into the anonymous class declared at line 23. ([squid:S3398](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS3398))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=293&issues=AWbRiJzSSgN4uCclDZzu&open=AWbRiJzSSgN4uCclDZzu)

[ISSUE_KEY]: AWbRiJzSSgN4uCclDZzu

saros-infrastructure commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SessionUtils.java:
```diff
@@ -0,0 +1,82 @@
+     */
+    public static boolean isShared(IResource resource) {
+
+        assert resource != null : "The given IResource must not be null.";
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Replace this assert with a proper check. ([squid:S4274](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS4274))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=293&issues=AWbRiJzSSgN4uCclDZzw&open=AWbRiJzSSgN4uCclDZzw)

[ISSUE_KEY]: AWbRiJzSSgN4uCclDZzw

saros-infrastructure commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SessionUtils.java:
```diff
@@ -0,0 +1,82 @@
+ * A utility class to provide easy access to functionality needing a session
+ * object.
+ */
+public class SessionUtils {
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Add a private constructor to hide the implicit public one. ([squid:S1118](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1118))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=293&issues=AWbRiJzSSgN4uCclDZzy&open=AWbRiJzSSgN4uCclDZzy)

[ISSUE_KEY]: AWbRiJzSSgN4uCclDZzy

saros-infrastructure commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SessionUtils.java:
```diff
@@ -0,0 +1,82 @@
+            });
+    }
+
+    private static void setSession(ISarosSession newSarosSession) {
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Move this method into the anonymous class declared at line 23. ([squid:S3398](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS3398))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=293&issues=AWbRiJzSSgN4uCclDZzv&open=AWbRiJzSSgN4uCclDZzv)

[ISSUE_KEY]: AWbRiJzSSgN4uCclDZzv

donut87 replied:
I have to agree with Sonar on this one. One line method used once inside the anonymous class, please inline this, or give a good reason not to.

tobous replied:
Done.

saros-infrastructure commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SessionUtils.java:
```diff
@@ -0,0 +1,82 @@
+     */
+    public static boolean isShared(SPath path) {
+
+        assert path != null : "The given SPath must not be null.";
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Replace this assert with a proper check. ([squid:S4274](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS4274))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=293&issues=AWbRiJzSSgN4uCclDZzx&open=AWbRiJzSSgN4uCclDZzx)

[ISSUE_KEY]: AWbRiJzSSgN4uCclDZzx

tobous commented:
I'm not sure about the sonar message concerning the asserts. I just tried to replicate the functionality of the IntelliJ [NotNull annotation](https://www.jetbrains.com/help/idea/nullable-and-notnull-annotations.html) (which I used before I decided to move this to the core).

saros-infrastructure commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SessionUtils.java:
```diff
@@ -0,0 +1,78 @@
+     */
+    public static boolean isShared(SPath path) {
+
+        assert path != null : "The given SPath must not be null.";
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Replace this assert with a proper check. ([squid:S4274](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS4274))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=293&issues=AWbRiJzSSgN4uCclDZzx&open=AWbRiJzSSgN4uCclDZzx)

[ISSUE_KEY]: AWbRiJzSSgN4uCclDZzx

saros-infrastructure commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SessionUtils.java:
```diff
@@ -0,0 +1,78 @@
+     */
+    public static boolean isShared(IResource resource) {
+
+        assert resource != null : "The given IResource must not be null.";
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Replace this assert with a proper check. ([squid:S4274](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS4274))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=293&issues=AWbRiJzSSgN4uCclDZzw&open=AWbRiJzSSgN4uCclDZzw)

[ISSUE_KEY]: AWbRiJzSSgN4uCclDZzw

tobous commented:
Could you elaborate why using the session manager (I assume you mean the lifecyclelistener to get the session object) should be avoided?

And no, this change is not necessarily necessary. I just noticed that in IntelliJ, almost all classes use the EditorManager to get session objects. (The EditorManager itself uses the lifecyclelistener.) This did not seem like a good usecase for the EditorManager. This especially bothered me when implementing #294, as it adds a lot of ISarosSession.isShared() checks (all of them are from classes belonging to the plugin scope and all of them would otherwise be using the EditorManager to get a session object).

So I decided to create this utility class. It basically encapsulates the same workflow as using the EditorManager and adds the needed nullcheck for the session object.

I would guess that the problem in the Saros/I part is that way to many components that belong inside the session context are still part of the plugin context, leading to such problems.

Do you have any suggestions on how to handle this better? Otherwise, should I just move this to the IntelliJ package and aim to remove it once the IntelliJ scopes are cleaned up (and maybe add a todo for that)?

tobous commented:
Moved class to IntelliJ package and added a note to the class javadoc explaining the context issue and a todo to remove the class once the context are separated correctly.

saros-infrastructure commented on de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/session/SessionUtils.java:
```diff
@@ -0,0 +1,88 @@
+     */
+    public static boolean isShared(IResource resource) {
+
+        assert resource != null : "The given IResource must not be null.";
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Replace this assert with a proper check. ([squid:S4274](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS4274))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=293&issues=AWbRiJzSSgN4uCclDZzw&open=AWbRiJzSSgN4uCclDZzw)

[ISSUE_KEY]: AWbRiJzSSgN4uCclDZzw

saros-infrastructure commented on de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/session/SessionUtils.java:
```diff
@@ -0,0 +1,88 @@
+     */
+    public static boolean isShared(SPath path) {
+
+        assert path != null : "The given SPath must not be null.";
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Replace this assert with a proper check. ([squid:S4274](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS4274))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=293&issues=AWbRiJzSSgN4uCclDZzx&open=AWbRiJzSSgN4uCclDZzx)

[ISSUE_KEY]: AWbRiJzSSgN4uCclDZzx

tobous commented:
Rebased on master and replaced asserts with IntelliJ NotNull annotations.

saros-infrastructure commented on de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/LocalEditorManipulator.java:
```diff
@@ -57,7 +60,7 @@ public void initialize(EditorManager editorManager) {
      */
     public Editor openEditor(SPath path, boolean activate) {
-        if (!manager.getSession().isShared(path.getResource())) {
+        if (!SessionUtils.isShared(path.getResource())) {
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this use of "SessionUtils"; it is deprecated. ([squid:CallToDeprecatedMethod](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=293&issues=AWcxWGhu0POp3Yiz0EjG&open=AWcxWGhu0POp3Yiz0EjG)

[ISSUE_KEY]: AWcxWGhu0POp3Yiz0EjG

saros-infrastructure replied:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this use of "SessionUtils"; it is deprecated. ([squid:CallToDeprecatedMethod](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=293&issues=AWcxWGhu0POp3Yiz0EjG&open=AWcxWGhu0POp3Yiz0EjG)

[ISSUE_KEY]: AWcxWGhu0POp3Yiz0EjG

saros-infrastructure replied:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this use of "SessionUtils"; it is deprecated. ([squid:CallToDeprecatedMethod](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=293&issues=AWcxWGhu0POp3Yiz0EjG&open=AWcxWGhu0POp3Yiz0EjG)

[ISSUE_KEY]: AWcxWGhu0POp3Yiz0EjG

saros-infrastructure commented on de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/session/SessionUtils.java:
```diff
@@ -0,0 +1,90 @@
+ */
+//TODO consider to remove this class after the session & plugin context have been correctly separated
+@Deprecated
+public class SessionUtils {
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Add the missing @deprecated Javadoc tag. ([squid:MissingDeprecatedCheck](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AMissingDeprecatedCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=293&issues=AWcxWGhR0POp3Yiz0EfN&open=AWcxWGhR0POp3Yiz0EfN)

[ISSUE_KEY]: AWcxWGhR0POp3Yiz0EfN

saros-infrastructure replied:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Do not forget to remove this deprecated code someday. ([squid:S1133](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1133))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=293&issues=AWcxWGhR0POp3Yiz0EfM&open=AWcxWGhR0POp3Yiz0EfM)

[ISSUE_KEY]: AWcxWGhR0POp3Yiz0EfM

saros-infrastructure commented on de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/editor/LocalEditorHandler.java:
```diff
@@ -156,7 +157,7 @@ private Editor openEditor(@NotNull VirtualFile virtualFile,
             return null;
 
-        }else if (!manager.getSession().isShared(path.getResource())) {
+        }else if (!SessionUtils.isShared(path)) {
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this use of "SessionUtils"; it is deprecated. ([squid:CallToDeprecatedMethod](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=293&issues=AWcxWGhf0POp3Yiz0EfP&open=AWcxWGhf0POp3Yiz0EfP)

[ISSUE_KEY]: AWcxWGhf0POp3Yiz0EfP

saros-infrastructure replied:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this use of "SessionUtils"; it is deprecated. ([squid:CallToDeprecatedMethod](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=293&issues=AWcxWGhf0POp3Yiz0EfP&open=AWcxWGhf0POp3Yiz0EfP)

[ISSUE_KEY]: AWcxWGhf0POp3Yiz0EfP

saros-infrastructure replied:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this use of "SessionUtils"; it is deprecated. ([squid:CallToDeprecatedMethod](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=293&issues=AWcxWGhf0POp3Yiz0EfP&open=AWcxWGhf0POp3Yiz0EfP)

[ISSUE_KEY]: AWcxWGhf0POp3Yiz0EfP

saros-infrastructure commented on de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/session/SessionUtils.java:
```diff
@@ -0,0 +1,93 @@
+ */
+//TODO consider to remove this class after the session & plugin context have been correctly separated
+@Deprecated
+public class SessionUtils {
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Do not forget to remove this deprecated code someday. ([squid:S1133](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1133))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=293&issues=AWcxWGhR0POp3Yiz0EfM&open=AWcxWGhR0POp3Yiz0EfM)

[ISSUE_KEY]: AWcxWGhR0POp3Yiz0EfM

saros-infrastructure replied:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Do not forget to remove this deprecated code someday. ([squid:S1133](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1133))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=293&issues=AWcxWGhR0POp3Yiz0EfM&open=AWcxWGhR0POp3Yiz0EfM)

[ISSUE_KEY]: AWcxWGhR0POp3Yiz0EfM

tobous commented on de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/session/SessionUtils.java:
```diff
@@ -0,0 +1,89 @@
+ * </p>
+ */
+//TODO consider to remove this class after the session & plugin context have been correctly separated
+public class SessionUtils {
```
Done.

tobous commented:
@m273d15 Just rebased (without any interaction) on the current master. Please re-approve.
